rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,136 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
             .await
-        {
-            Ok(records) => records,
-            Err(e) => {
-                return Err(MergeMetadataResultsOperatorError::LogMaterializationError(
-                    e,
-                ));
+            .map_err(|e| {
+                tracing::error!("Error materializing log: {}", e);
+                MergeMetadataResultsOperatorError::LogMaterializationError(e)
+            })?;
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        let merged_oids_holder: Vec<u32>;
+        let merged_offset_ids = match (&input.filtered_offset_ids, &input.user_offset_ids) {
+            (Some(fids), Some(uids)) => {
+                merged_oids_holder = merge_sorted_vecs_conjunction(fids, uids);
+                &merged_oids_holder
+            }
+            (Some(oids), None) | (None, Some(oids)) => oids,
+            // If both filter and user offset ids are None, it suggests user did not specify any filter. We fetch all offset ids using the record segment reader
+            _ => {
+                let live_log_offset_ids = mat_records
+                    .iter()
+                    .filter_map(|(log, _)| {
+                        (log.final_operation != MaterializedLogOperation::DeleteExisting)
+                            .then_some(log.offset_id)
+                    })
+                    .collect::<BTreeSet<_>>()
+                    .into_iter()
+                    .collect::<Vec<_>>();
+                merged_oids_holder = match &record_segment_reader {
+                    Some(reader) => {
+                        let compact_offset_ids =
+                            reader.get_all_offset_ids().await.map_err(|e| {
+                                tracing::error!("Error reading record segment: {}", e);
+                                MergeMetadataResultsOperatorError::RecordSegmentReadError
+                            })?;
+                        merge_sorted_vecs_disjunction(&live_log_offset_ids, &compact_offset_ids)
+                    }
+                    None => live_log_offset_ids,
+                };
+                &merged_oids_holder
             }
         };
 
-        // Step 2: Hydrate data.
-        let mut ids: Vec<String> = Vec::new();
+        // Truncate the offset ids using offset and limit
+        let skip_count = input.offset.map(|o| o as usize).unwrap_or(0);
+        let take_count = input
+            .limit
+            .map(|l| l as usize)
+            .unwrap_or(merged_offset_ids.len() - skip_count);
+
+        // Hydrate data
+        let truncated_offset_ids = merged_offset_ids[skip_count..(skip_count + take_count)].iter();
+        let truncated_offset_id_order: HashMap<u32, usize> = truncated_offset_ids
+            .clone()
+            .enumerate()
+            .map(|(i, offset_id)| (*offset_id, i))
+            .collect();
+        let mut ids: Vec<String> = vec![String::new(); take_count];
```

Comment:
this need not necessarily be `take_count` right? For e.g. if merged_offset_ids has a length of 3 and user specifies offset of 2 and limit of 2

Reply:
`take_count` should be the exact size of the returned result. In your case the actual `take_count` should be capped to `3 [merged_offset_id.len()] - 2 [offset] = 1`. Please take a look at the next PR for the fix for `limit` and `offset`.

Reply:
I'll include the fix for `limit` and `offset` to the current PR.